apps/omg/lib/omg/ethereum_client_monitor.ex

```diff
@@ -14,50 +14,133 @@
+  When the process is started we immediately make an RPC call to retrieve the height, we proceed to open a subscription towards the client.
+  If the client connection drops, we get notified (`def handle_info({:EXIT, _from, _}...`) and raise an alarm and proceed with periodical health checks.
+  A health check makes an RPC call and checks for correct response (is_number) - if that succeeds, there's a high probability websocket connection subscription will work as well.
+
+  The implementation assumes ws subscription to the Ethereum client continues to work indefinitely. We'll see how that works in practice.
   """
   use GenServer
   require Logger
   alias OMG.Eth
+  alias OMG.Eth.Encoding
+  alias OMG.Eth.SubscriptionWorker
 
   @default_interval Application.get_env(:omg, :client_monitor_interval_ms)
   @type t :: %__MODULE__{
           interval: pos_integer(),
           tref: reference() | nil,
           alarm_module: module(),
-          raised: boolean()
+          raised: boolean(),
+          ethereum_height: integer | :error,
+          ws_url: binary() | nil
         }
-  defstruct interval: @default_interval, tref: nil, alarm_module: nil, raised: true
+  defstruct interval: @default_interval,
+            tref: nil,
+            alarm_module: nil,
+            raised: true,
+            ethereum_height: :error,
+            ws_url: nil
+
+  def get_ethereum_height do
+    GenServer.call(__MODULE__, :ethereum_height)
+  end
 
   def start_link(args) do
     GenServer.start_link(__MODULE__, args, name: __MODULE__)
   end
 
-  def init([alarm_module]) do
+  def init([_ | _] = opts) do
+    alarm_module = Keyword.get(opts, :alarm_module)
+    false = Process.flag(:trap_exit, true)
     _ = Logger.info("Starting Ethereum client monitor.")
-    install()
-    state = %__MODULE__{alarm_module: alarm_module}
+    install_alarm_handler()
+    ethereum_height = check()
+
+    state = %__MODULE__{
+      alarm_module: alarm_module,
+      ethereum_height: ethereum_height,
+      ws_url: Keyword.get(opts, :ws_url)
+    }
+
     _ = alarm_module.set({:ethereum_client_connection, Node.self(), __MODULE__})
-    _ = raise_clear(alarm_module, state.raised, check())
-    {:ok, tref} = :timer.send_after(state.interval, :health_check)
-    {:ok, %{state | tref: tref}}
+    _ = raise_clear(alarm_module, state.raised, ethereum_height)
+    {:ok, state, {:continue, :ws_connect}}
   end
 
-  # gen_event
+  # gen_event init
   def init(_args) do
     {:ok, %{}}
   end
 
-  def handle_info(:health_check, state) do
-    _ = raise_clear(state.alarm_module, state.raised, check())
+  def handle_continue(:ws_connect, state) do
+    _ = Logger.debug("Ethereum client monitor starting a WS newHeads subscription.")
+
+    params =
+      case state.ws_url do
+        nil -> [listen_to: "newHeads"]
+        ws_url -> [listen_to: "newHeads", ws_url: ws_url]
+      end
+
+    _ = SubscriptionWorker.start_link(params)
```

Comment:
An Elixir question for you @InoMurko - I notice that we often match on `_` when we don't care about the return value at all. Would it be the same to just call the function without the match or is there some other benefit to this?

Reply:
Essentially the reason is Dialyzer and decision making:
Dialyzer would throw a tantrum. Hey, you're ignoring a return value from a function call that *might* be important. So we're matching it with *_* so that we make a conscious decision you really don't care about the return.